src/request.ts

```diff
@@ -0,0 +1,145 @@
+
+    if (method !== 'GET') {
+        fetchHeaders = {
+            'Content-Type': 'application/json',
+            ...headers,
+        };
+    }
+
+    const fetchOptions = {
+        method,
+        mode,
+        headers: fetchHeaders,
+        credentials: mode === 'cors' ? 'include' : 'same-origin',
+        ...options,
+    } as RequestInit;
+
+    return fetch(url, fetchOptions)
+        .then(checkStatus)
+        .then((res: Response) => {
+            let resJSON = {};
+
+            if ('json' in res) {
+                resJSON = res.json();
+            }
+
+            return resJSON;
+        });
+};
+
+export interface ArgumentErrors {
+    [key: string]: [string];
+}
+
+export interface ParsedResponseError {
+    error: string;
+    description: string;
+    argumentErrors: ArgumentErrors;
+}
+
+export interface JSONResponseData {
+    error?: string;
+    error_description?: string;
+    error_detail?: {
+        ARGUMENTS_ERROR?: ArgumentErrors;
+        [propName: string]: any;
+    };
+}
+
+const hasArgumentsError = (responseData: JSONResponseData): boolean =>
+    isObject(responseData['error_detail']) &&
+    isObject(responseData['error_detail']['ARGUMENTS_ERROR']);
+
+/**
+ * Parse v3 errors into an array of objects representing the errors returned by
+ * the API. The format of the parsed errors looks like:
+ *
+ *  {
+ *      status_code: 400,
+ *      error: 'ERROR_CODE',
+ *      description: 'Description of the error
+ *  }
+ *
+ * An ARGUMENTS_ERROR looks like:
+ *
+ *  {
+ *      error: 'ARGUMENTS_ERROR',
+ *      description: 'Some of the fields were invalid or something',
+ *      argumentErrors: {
+ *          attr1: ['INVALID'],
+ *          attr2: ['This field is required']
+ *      }
+ *  }
+ *
+ */
+export const parseError = (
+    responseData: JSONResponseData
+): ParsedResponseError => {
+    if (!responseData.error) {
+        // Weird error format, return null
+        return null;
+    }
+
+    let error = {
+        error: responseData.error,
+        description: responseData['error_description'],
+    } as ParsedResponseError;
+
+    if (hasArgumentsError(responseData)) {
+        error.argumentErrors = responseData['error_detail']['ARGUMENTS_ERROR'];
+    }
+
+    return error;
+};
+
+/**
+ * Designed to work with `checkStatus`, or any function that
+ * raises an error on an invalid status. The error raised should have a `response`
+ * property with the original response object.
+ *
+ * Example usage:
+ *
+ * fetchJSON('/api/v3/test/path', {'body': someData})
+ *     .catch(catchStatusError)
+ *     .catch(({response, parsedError}) => doSomethingOnError())
+ *     .then(doSomethingOnSuccess);
+ */
+export const catchStatusError = (response: Response): Promise<any> =>
+    new Promise((resolve, reject) => {
+        response
+            .json()
+            .then((responseData) => parseError(responseData))
+            .then((parsedError) => reject({response, parsedError}))
```

Comment:
Is there a reason return both response and parsedError? can you do response: parsedError instead?

Reply:
I'm assuming so that you can have access to the raw `response` as well as the `parsedError`. The `response` will have other meta data (like response code) that _may_ be needed in addition to the parsed error.

This is also how the code currently is in core.